internal/events/aggregator_rewind.go

```diff
@@ -0,0 +1,284 @@
+}
+
+func newRewinder(ag *aggregator) *rewinder {
+	return &rewinder{
+		ctx:              log.WithLogField(ag.ctx, "role", "aggregator-rewind"),
+		aggregator:       ag,
+		database:         ag.database,
+		data:             ag.data,
+		retry:            ag.retry,
+		loop1Done:        make(chan struct{}),
+		loop2Done:        make(chan struct{}),
+		rewindRequests:   make(chan rewind, config.GetInt(config.EventAggregatorRewindQueueLength)),
+		loop2ShoulderTap: make(chan bool, 1),
+		minRewindTimeout: config.GetDuration(config.EventAggregatorRewindTimeout),
+		readyRewinds:     make(map[fftypes.UUID]bool),
+	}
+}
+
+func (rw *rewinder) start() {
+	go rw.rewindLoop1()
+	go rw.rewindLoop2()
+}
+
+// rewindLoop1 is responsible from taking requests to the aggregator for rewinds, from other
+// event loops that execute completely independently.
+// It takes them in as fast as possible, and puts them in a staging area ready for the aggregator
+// itself to mark it's finished processing anything in-flight
+func (rw *rewinder) rewindLoop1() {
+	defer close(rw.loop1Done)
+
+	for {
+		select {
+		case rewind := <-rw.rewindRequests:
+			rw.mux.Lock()
+			rw.queuedRewinds = append(rw.queuedRewinds, &rewind)
+			rw.mux.Unlock()
+
+			// Shoulder tap at this point, to get the event loop to pop and tell us
+			// we can move the queued rewinds to staged
+			rw.aggregator.eventPoller.shoulderTap()
+		case <-rw.ctx.Done():
+			log.L(rw.ctx).Debugf("Rewind loop 1 stopping")
+			return
+		}
+	}
+}
+
+// rewindLoop2 does the heavy lifting of taking rewinds that the aggregator has marked staged,
+// and doing the DB queries required to find out what BatchIDs need to be resolved.
+// These then go to the readyRewinds map, ready for the aggregator to pop them
+func (rw *rewinder) rewindLoop2() {
+	defer close(rw.loop2Done)
+
+	var timerCtx context.Context
+	var timerCancel context.CancelFunc
+	for {
+		var popCtx context.Context
+		if timerCtx != nil {
+			// Pop when we've timed out
+			popCtx = timerCtx
+		} else {
+			// Only pop for new work, or shutdown
+			popCtx = rw.ctx
+		}
+		select {
+		case <-popCtx.Done():
+			if timerCtx == nil && timerCancel == nil {
+				log.L(rw.ctx).Debugf("Rewind loop 2 stopping")
+				return
+			}
+			timerCancel()
+			timerCtx = nil
+			timerCancel = nil
+			if rw.processStagedRewinds() {
+				// Shoulder tap at this point, to get the event loop to pop and retrieve
+				// the rewinds we have just derived batch IDs from
+				rw.aggregator.eventPoller.shoulderTap()
+			}
+		case <-rw.loop2ShoulderTap:
+			if timerCtx == nil {
+				// We've been told at least one rewind is available, so start a timer
+				// to wait for more to accumulate before we un-stage them
+				timerCtx, timerCancel = context.WithTimeout(rw.ctx, rw.minRewindTimeout)
+			}
+		}
+	}
+}
+
+func (rw *rewinder) processStagedRewinds() bool {
+	batchIDs := make(map[fftypes.UUID]bool)
+	var msgRewinds []*fftypes.UUID
+	var newBlobHashes []driver.Value
```

Comment:
any reason not to type this to `*fftypes.Bytes32`?

Reply:
aha, it's valid in `sq.Eq`, but not in our builder `In("blob.hash", newHashes)` - so while I can remove the ones below on the `Database` package inputs, I can't remove this one without significant extra work.